### PR TITLE
Delete name index provided

### LINKED ISSUE #143: Delete Command - Throw an error if both index and name is provided
parse once for name
parse once for index
if both found, throw error

### PR DESCRIPTION
**Pull request by wapisai**
# This PR addresses the issues #143 
fix #143 
This Pull Request enhances the parsing of Delete command - An error will be thrown if both index and name fields are provided
Note: 
`delete INDEX` or `delete n/NAME` will result in correct delete action
`delete 1 n/Joe` (i.e index followed by name) will result in "Please provide either an index or a name, not both."
`delete n/Joe 1` (i.e. name followed by "index") will result in "The client name provided is invalid" as the "index" is considered part of the name
`delete` will result in "Please provide either an index or a name."

### Changes Made:
Delete Command Enhancement: An error will be thrown if both index and name are provided
Unit Test Expansion: Introduced comprehensive unit tests in DeleteCommandTest and DeleteCommandParserTest to rigorously assess the new functionalities. These tests cover a wide array of scenarios, including both edge cases and common use patterns, to ensure the feature's reliability.

### Testing (YET TO DO):
Unit Testing: Developed and executed a suite of new unit tests designed to rigorously validate the delete operation across various scenarios, emphasizing error handling and behavior correctness.
Manual Testing: Performed extensive manual testing to guarantee that the integration of this new feature does not disrupt existing functionalities, ensuring a seamless user experience.
Code Quality and Compatibility:

### Compatibility Assurance: 
Verified the new feature's compatibility with existing commands and functionalities through detailed automated regression testing, ensuring no disruptions to the current system.

### Code Quality Maintenance: 
Conducted thorough code quality checks to uphold our commitment to high standards and adherence to the project's coding conventions.

### Additional Notes:
Priority for this feature has been set to Medium, reflecting its importance to user experience.
We encourage further review and testing from the community to refine and validate the new tests and functionalities introduced.

### Please feel free to provide feedback or suggestions for further improvements!

### CODE REVIEW

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 1]
```diff
@@ -12,21 +12,45 @@
...
+            name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
+            trimmed = args.replace("n/" + name.fullName, "");
         }
 
+        // Parse index if present
+        Index index;
+        try {
+            index = ParserUtil.parseIndex(trimmed);
+        } catch (ParseException pe) {
+            index = null;
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:39]: not too sure about this, it looks like exceptions are used as control flow. anyone else agree?

Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:39]: I think its ok following from `EditCommandParser`

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 2]
```diff
@@ -12,21 +12,45 @@
...
     /**
      * Parses the given {@code String} of arguments in the context of the DeleteCommand
      * and returns a DeleteCommand object for execution.
      * @throws ParseException if the user input does not conform the expected format
      */
     public DeleteCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME);
+        String trimmed = args;
+
+        // Parse name if present
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:27]: I think should add `argMultimap.verifyNoDuplicatePrefixesFor` check for double `n/`

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 3]
```diff
@@ -12,21 +12,45 @@
...
+        Name name;
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:28]:
```suggestion
        Name name = null;
```

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java (hunk 1 above)
Comment (4llysa) [reply src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:39]: this is very similar to udhaya's pr for #140 and my pr for #154. I was thinking about it but seems that to check whether `trimmed` can be parsed, we need to parse it first

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 4]
```diff
@@ -12,21 +12,45 @@
...
+        try {
+            index = ParserUtil.parseIndex(trimmed);
+        } catch (ParseException pe) {
+            index = null;
+        }
+
+        DeleteCommand deleteCommand;
+        if (namePrefixPresent(argMultimap) && index == null) {
+            name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
+            deleteCommand = new DeleteCommand(name);
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:45]: Name shld be initialised alr
```suggestion
        if (namePrefixPresent(argMultimap) && index == null) {
            assert name != null //remove this if the checks fail or smth codecov might be problem
            deleteCommand = new DeleteCommand(name);
```

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 5]
```diff
@@ -12,21 +12,45 @@
...
+        } else if (!namePrefixPresent(argMultimap) && index != null) {
+            deleteCommand = new DeleteCommand(index);
+        } else if (namePrefixPresent(argMultimap) && index != null) {
+            throw new ParseException(INDEX_AND_NAME_PROVIDED);
+        } else {
+            throw new ParseException(NO_FIELDS_PROVIDED);
+        }
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:52]:
```suggestion
        } else if (namePrefixPresent(argMultimap) && index != null) {
            throw new ParseException(INDEX_AND_NAME_PROVIDED + "\n" + DeleteCommand.MESSAGE_DELETE_HELP);
        } else {
            throw new ParseException(NO_FIELDS_PROVIDED + "\n" + DeleteCommand.MESSAGE_DELETE_HELP);
        }
```

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 6]
```diff
@@ -12,21 +12,45 @@
  */
 public class DeleteCommandParser implements Parser<DeleteCommand> {
 
+    public static final String INDEX_AND_NAME_PROVIDED = "Please provide either an index or a name, not both.";
+    public static final String NO_FIELDS_PROVIDED = "Please provide either an index or a name.";
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:16]: Move these two to `DeleteCommand`

```suggestion
    public static final String MESSAGE_INDEX_AND_NAME_PROVIDED = "Please provide either an index or a name, not both.";
    public static final String MESSAGE_NO_FIELDS_PROVIDED = "Please provide either an index or a name.";
```

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
LGTM but minor changes

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java (hunk 2 above)
Comment (wapisai) [reply src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:27]: Resolved thanks!

**Comment by wapisai** [comment]
Summary of delete error messages
`delete 1 n/name` - (i.e. valid index AND valid name) -> "Please provide either an index or a name, not both."
`delete n/name 1` - (i.e. valid name AND valid index) -> "The client name provided is invalid" (index is considered part of client name"
`delete -1` - (i.e. invalid index) -> "Index is not a non-zero unsigned integer."
`delete n/wrongname` (i.e. invalid name) -> "The client name provided is invalid"

`delete 1 1` - (i.e. multiple valid/invalid integers) -> "Index is not a non-zero unsigned integer."
`delete n/name n/invalidname` - (multiple valid/invalid name) -> "Multiple values specified for the following single-valued field(s): n/"

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 7]
```diff
@@ -19,14 +21,31 @@ public class DeleteCommandParser implements Parser<DeleteCommand> {
...
+            assert name != null;
+            deleteCommand = new DeleteCommand(name);
+        } else if (!namePrefixPresent(argMultimap) && index != null) {
+            deleteCommand = new DeleteCommand(index);
+        } else if (namePrefixPresent(argMultimap) && index != null) {
+            throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
+        } else {
+            throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
+        }
+        return deleteCommand;
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:48]:
```suggestion
        Index index = null;
        String checkPreamble = argMultimap.getPreamble();
        if (!checkPreamble.isBlank()) {
            index = ParserUtil.parseIndex(trimmed);
        }



        DeleteCommand deleteCommand;
        if (namePrefixPresent(argMultimap) && checkPreamble.isBlank()) {
            assert name != null;
            deleteCommand = new DeleteCommand(name);
        } else if (!namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
            deleteCommand = new DeleteCommand(index);
        } else if (namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
            throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
        } else {
            throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
        }
        return deleteCommand;
```

File: src/main/java/seedu/realodex/logic/parser/ParserUtil.java [hunk 8]
```diff
@@ -32,6 +32,9 @@ public class ParserUtil {
      */
     public static Index parseIndex(String oneBasedIndex) throws ParseException {
         String trimmedIndex = oneBasedIndex.trim();
+        if (trimmedIndex.isEmpty()) {
+            return null;
+        }
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/ParserUtil.java:37]:
```suggestion
```

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
Try these changes

**Review by wapisai — COMMENTED** [review]
Thanks!

File: src/test/java/seedu/realodex/logic/parser/DeleteCommandParserTest.java [hunk 9]
```diff
@@ -44,4 +52,22 @@ public void parse_invalidName_throwParseException() {
...
     }
+
+    @Test
+    public void parse_nameAndIndex_throwsParseException() {
+        assertParseFailure(parser, "1 n/James", String.format(DeleteCommand.MESSAGE_INDEX_AND_NAME_PROVIDED));
+    }
+
+    @Test
+    public void parse_noFields_throwsParseException() {
+        assertParseFailure(parser, "", String.format(MESSAGE_NO_FIELDS_PROVIDED));
```
Comment (jyztintan) [inline src/test/java/seedu/realodex/logic/parser/DeleteCommandParserTest.java:63]: nice!

**Review by jyztintan — APPROVED** [review]
Thanks for fixing this bug 🫡

Thoughtful and meaningful tests added that makes good use of equivalence partitioning

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 10]
```diff
@@ -19,14 +21,36 @@ public class DeleteCommandParser implements Parser<DeleteCommand> {
...
+        if (namePrefixPresent(argMultimap) && checkPreamble.isBlank()) {
+            assert name != null;
+            deleteCommand = new DeleteCommand(name);
+        } else if (!namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
+            deleteCommand = new DeleteCommand(index);
+        } else if (namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
+            throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
+        } else {
+            throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
+        }
```
[the remark this answers is not in the document]
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:52]: Agree consider doing another class like `CreateDeleteCommand` or method

```suggestion
public DeleteCommand parse(String args) throws ParseException {
    ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME);
    argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME);
    String trimmed = args;

    // Parse name if present
    Name name = null;
    if (namePrefixPresent(argMultimap)) {
        name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
        trimmed = args.replace("n/" + name.fullName, "");
    }

    Index index = null;
    String checkPreamble = argMultimap.getPreamble();
    if (!checkPreamble.isBlank()) {
        index = ParserUtil.parseIndex(trimmed);
    }

    return createDeleteCommand(argMultimap, name, index, checkPreamble);
}

private DeleteCommand createDeleteCommand(ArgumentMultimap argMultimap, Name name, Index index, String checkPreamble)
        throws ParseException {
    if (namePrefixPresent(argMultimap) && checkPreamble.isBlank()) {
        assert name != null;
        return new DeleteCommand(name);
    } else if (!namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
        return new DeleteCommand(index);
    } else if (namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
        throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
    } else {
        throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
    }
}

```

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
LGTM just slap issues

**Comment by UdhayaShan1** [comment]
## Please pull the PR to your local repo then make the changes, the suggestions are not meant to be committed on the webpage itself. You cant check whether the code compiles or not on the page

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
Please pull the PR to your local repo then make the changes, the suggestions are not meant to be committed on the webpage itself. You cant check whether the code compiles or not on the page

**Review by appleraincoat — COMMENTED** [review]
The code currently does not compile, please pull locally on your computer and fix the issues and test before pushing again. Otherwise, the rest LGTM.

File: src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java [hunk 11]
```diff
@@ -19,14 +21,34 @@ public class DeleteCommandParser implements Parser<DeleteCommand> {
...
-            Index index = ParserUtil.parseIndex(args);
+        } else if (!namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
             return new DeleteCommand(index);
+        } else if (namePrefixPresent(argMultimap) && !checkPreamble.isBlank()) {
+            throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
+        } else {
+            throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
         }
-
     }
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/DeleteCommandParser.java:52]: ChatGPTed refinement of the class, just to ensure SLAP


```suggestion
    package seedu.realodex.logic.parser;

import static seedu.realodex.logic.commands.DeleteCommand.MESSAGE_INDEX_AND_NAME_PROVIDED;
import static seedu.realodex.logic.commands.DeleteCommand.MESSAGE_NO_FIELDS_PROVIDED;
import static seedu.realodex.logic.parser.CliSyntax.PREFIX_NAME;

import seedu.realodex.commons.core.index.Index;
import seedu.realodex.logic.commands.DeleteCommand;
import seedu.realodex.logic.parser.exceptions.ParseException;
import seedu.realodex.model.person.Name;

/**
 * Parses input arguments and creates a new DeleteCommand object
 */
public class DeleteCommandParser implements Parser<DeleteCommand> {

    public DeleteCommand parse(String args) throws ParseException {
        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME);
        String trimmedArgs = args;

        Name name = parseName(argMultimap);
        Index index = parseIndex(argMultimap);

        return createDeleteCommand(name, index);
    }

    private Name parseName(ArgumentMultimap argMultimap) throws ParseException {
        if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
            String fullName = argMultimap.getValue(PREFIX_NAME).get();
            return ParserUtil.parseName(fullName);
        }
        return null;
    }

    private Index parseIndex(ArgumentMultimap argMultimap) throws ParseException {
        String preamble = argMultimap.getPreamble();
        if (!preamble.isBlank()) {
            return ParserUtil.parseIndex(preamble);
        }
        return null;
    }

    private DeleteCommand createDeleteCommand(Name name, Index index) throws ParseException {
        if (name != null && index != null) {
            throw new ParseException(MESSAGE_INDEX_AND_NAME_PROVIDED);
        } else if (name != null) {
            return new DeleteCommand(name);
        } else if (index != null) {
            return new DeleteCommand(index);
        } else {
            throw new ParseException(MESSAGE_NO_FIELDS_PROVIDED);
        }
    }
}

```

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
chatgpted refinement

**Review by UdhayaShan1 — APPROVED** [review]
